src/mesh/Draco_Mesh.cc

```diff
@@ -0,0 +1,265 @@
+//----------------------------------*-C++-*----------------------------------//
+/*!
+ * \file   mesh/Draco_Mesh.cc
+ * \date   May 2018
+ * \brief  Draco_Mesh class implementation file.
+ * \note   Copyright (C) 2018 Los Alamos National Security, LLC.
+ *         All rights reserved. */
+//---------------------------------------------------------------------------//
+
+#include "Draco_Mesh.hh"
+#include "ds++/Assert.hh"
+#include <algorithm>
+#include <iostream>
+#include <numeric>
+
+namespace rtt_mesh {
+
+//---------------------------------------------------------------------------//
+// CONSTRUCTOR
+//---------------------------------------------------------------------------//
+/*!
+ * \brief Draco_Mesh constructor.
+ *
+ * \param[in] dimension_ dimension of mesh
+ * \param[in] geometry_ enumerator of possible coordinate system geometries
+ * \param[in] cell_type_ number of vertices for each cell
+ * \param[in] cell_to_node_linkage_ serialized map of cell indices to node
+ * indices.
+ * \param[in] side_set_flag_ map of side indices (per cell) to side flag (global
+ * index for a side).
+ * \param[in] side_node_count_ number of nodes per each cell on a side of
+ * the mesh.
+ * \param[in] side_to_node_linkage_ serialized map of side indices (per side
+ * cell) to node indices.
+ * \param[in] coordinates_ serialized map of node index to coordinate values.
+ * \param[in] global_node_number_ map of local to global node index (vector
+ * subscript is local node index and value is global node index; for one
+ * process, this is the identity map).
+ */
+Draco_Mesh::Draco_Mesh(unsigned dimension_, Geometry geometry_,
+                       const std::vector<unsigned> &cell_type_,
+                       const std::vector<unsigned> &cell_to_node_linkage_,
+                       const std::vector<unsigned> &side_set_flag_,
+                       const std::vector<unsigned> &side_node_count_,
+                       const std::vector<unsigned> &side_to_node_linkage_,
+                       const std::vector<double> &coordinates_,
+                       const std::vector<unsigned> &global_node_number_)
+    : dimension(dimension_), geometry(geometry_), num_cells(cell_type_.size()),
+      num_nodes(global_node_number_.size()), side_set_flag(side_set_flag_) {
+
+  // Require(dimension_ <= 3);
+  // TODO: generalize mesh generation to 1D,3D (and uncomment requirment above)
+  Require(dimension_ == 2);
+
+  // build the layout (or linkage) of the mesh
+  compute_cell_to_cell_linkage(cell_type_, cell_to_node_linkage_,
+                               side_node_count_, side_to_node_linkage_);
```

Comment:
This linkage function may be possible to make `const` if the layout data moves to a separate class.  It may be useful to identify other requirements to motivate or deter this refactor as well.